File: JavaScript/프로그래머스/3/43163. 단어 변환/단어 변환.js
function solution(begin, target, words) {
    const queue = [{ word: begin, count: 0 }];
    const visited = new Array(words.length).fill(false);
    
    while (queue.length > 0) {
        const { word, count } = queue.shift();
        
        if (word === target) return count;
        
        for (let i = 0; i < words.length; i++) {
            if (!visited[i] && getDiffCount(word, words[i]) === 1) {
                visited[i] = true;
                queue.push({ word: words[i], count: count + 1 });
            }
        }
    }
    
    return 0;
}

function getDiffCount(word1, word2) {
    let diff = 0;
    for (let i = 0; i < word1.length; i++) {
        if (word1[i] !== word2[i]) diff++;
    }
    return diff;
}

File: Python3/프로그래머스/3/43163. 단어 변환/단어 변환.py
from collections import deque

def solution(begin, target, words):
    
    if target not in words:
        return 0

    def one_letter_diff(a, b):
        diff = 0
        for x, y in zip(a, b):
            if x != y:
                diff += 1
        return diff == 1

    q = deque()
    q.append((begin, 0))

    visited = set()

    while q:
        cur, steps = q.popleft()

        if cur == target:
            return steps

        for w in words:
            if w not in visited and one_letter_diff(cur, w):
                visited.add(w)
                q.append((w, steps + 1))
    return 0